pkgs/top-level/variants.nix

```diff
@@ -95,6 +95,36 @@ self: super: {
     };
   });
 
+  # Full package set with cuda on rocm off
+  # Mostly useful for asserting pkgs.pkgsCuda.torchWithCuda == pkgs.torchWithCuda and similar
+  pkgsCuda = nixpkgsFun {
+    config = super.config // {
+      cudaSupport = true;
+      rocmSupport = false;
+    };
+  };
+
+  # `pkgsForCudaArch` maps each CUDA capability in _cuda.db.cudaCapabilityToInfo to a Nixpkgs variant configured for
+  # that target system. For example, `pkgsForCudaArch.sm_90a.python3Packages.torch` refers to PyTorch built for the
+  # Hopper architecture, leveraging architecture-specific features.
+  # NOTE: Not every package set is supported on every architecture!
+  # See `Using pkgsForCudaArch` in doc/languages-frameworks/cuda.section.md for more information.
+  pkgsForCudaArch = lib.listToAttrs (
+    lib.map (cudaCapability: {
+      name = self._cuda.lib.mkRealArchitecture cudaCapability;
+      value = nixpkgsFun {
+        config = super.config // {
+          cudaSupport = true;
+          rocmSupport = false;
+          # Not supported by architecture-specific feature sets, so disable for all.
+          # Users can choose to build for family-specific feature sets if they wish.
+          cudaForwardCompat = false;
+          cudaCapabilities = [ cudaCapability ];
+        };
+      };
+    }) (lib.attrNames self._cuda.db.cudaCapabilityToInfo)
+  );
+
```

Comment:
FTR we'd discussed this with Connor, and it is a known concern that we here are bending the Nixpkgs' "public API" to compensate for Nix's CLI UX issues. This is somewhat evidenced by there being no in-tree consumers for selecting specific CUDA architectures. Still the intuitive way to achieve the desired UX "now", and I can't really judge what's worth what...

As such, even today this UX kinda could be implemented in a separate layer. Like we could literally write a builder-pattern wrapper for nixpkgs going like `(foo.buildPlatform.x86_64-linux.hostPlatform.aarch64-linux.cuda."86".withOverlay (fi: pre: ...)).result`, and a cli tool around such wrapper.